In [3]:
#Load the Packages Needed
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:
#Load the Data
pwt = pd.read_csv('https://raw.githubusercontent.com/jivizcaino/PWT_10.0/main/pwt100.csv', encoding="latin-1")
pwt

countrycode   country   currency_unit  year        rgdpe        rgdpo  \
0             ABW     Aruba  Aruban Guilder  1950          NaN          NaN   
1             ABW     Aruba  Aruban Guilder  1951          NaN          NaN   
2             ABW     Aruba  Aruban Guilder  1952          NaN          NaN   
3             ABW     Aruba  Aruban Guilder  1953          NaN          NaN   
4             ABW     Aruba  Aruban Guilder  1954          NaN          NaN   
...           ...       ...             ...   ...          ...          ...   
12805         ZWE  Zimbabwe       US Dollar  2015  40141.61719  39798.64453   
12806         ZWE  Zimbabwe       US Dollar  2016  41875.20313  40963.19141   
12807         ZWE  Zimbabwe       US Dollar  2017  44672.17578  44316.74219   
12808         ZWE  Zimbabwe       US Dollar  2018  44325.10938  43420.89844   
12809         ZWE  Zimbabwe       US Dollar  2019  42296.06250  40826.57031   

             pop       emp  avh        hc  ...     csh_x     csh_m     csh_r  \
0            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
1            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
2            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
3            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
4            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
...          ...       ...  ...       ...  ...       ...       ...       ...   
12805  13.814629  6.393752  NaN  2.584653  ...  0.140172 -0.287693 -0.051930   
12806  14.030331  6.504374  NaN  2.616257  ...  0.131920 -0.251232 -0.016258   
12807  14.236595  6.611773  NaN  2.648248  ...  0.126722 -0.202827 -0.039897   
12808  14.438802  6.714952  NaN  2.680630  ...  0.144485 -0.263658 -0.020791   
12809  14.645468  6.831017  NaN  2.713408  ...  0.213562 -0.270959 -0.089798   

           pl_c      pl_i      pl_g      pl_x      pl_m      pl_n      pl_k  
0           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
1           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
2           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
3           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
4           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
...         ...       ...       ...       ...       ...       ...       ...  
12805  0.479228  0.651287  0.541446  0.616689  0.533235  0.422764  1.533909  
12806  0.470640  0.651027  0.539631  0.619789  0.519718  0.416510  1.491724  
12807  0.473560  0.639560  0.519956  0.619739  0.552042  0.415592  1.514525  
12808  0.543757  0.655473  0.529867  0.641361  0.561526  0.425143  1.590120  
12809  0.494755  0.652439  0.500927  0.487763  0.430082  0.420675  1.384068  

[12810 rows x 52 columns]

In [5]:
#Prepare the Dataset and Clean the Data
pwt_subset = pwt[['country','countrycode','year','cgdpo','pop','emp','avh','hc','cn','labsh','ctfp']]
pwt_subset = pwt_subset[pwt_subset['year'] > 2009]
pwt_subset = pwt_subset.set_index('country')
pwt_subset = pwt_subset.dropna()
pwt_subset

countrycode  year        cgdpo        pop        emp  \
country                                                             
Argentina            ARG  2010  768478.2500  40.895752  18.076746   
Argentina            ARG  2011  868765.0625  41.320500  18.488029   
Argentina            ARG  2012  882438.8125  41.755196  18.791292   
Argentina            ARG  2013  885112.9375  42.196030  19.124418   
Argentina            ARG  2014  864304.0000  42.637511  19.383051   
...                  ...   ...          ...        ...        ...   
South Africa         ZAF  2014  699348.0625  54.544186  16.992683   
South Africa         ZAF  2016  710312.5000  56.207646  17.984968   
South Africa         ZAF  2017  726021.0000  57.009756  18.302843   
South Africa         ZAF  2018  733319.3750  57.792518  18.686357   
South Africa         ZAF  2019  735067.0625  58.558270  18.642710   

                      avh        hc          cn     labsh      ctfp  
country                                                              
Argentina     1760.224782  2.830078  2876207.25  0.440692  0.777526  
Argentina     1760.230369  2.858542  3099233.75  0.473124  0.828393  
Argentina     1736.103360  2.887292  3156028.75  0.520179  0.819774  
Argentina     1725.873553  2.916332  3166597.50  0.543918  0.813449  
Argentina     1705.428368  2.945663  3204576.50  0.543918  0.781934  
...                   ...       ...         ...       ...       ...  
South Africa  2209.091467  2.675951  2727571.50  0.552341  0.596181  
South Africa  2218.455068  2.762902  2773613.50  0.568343  0.566500  
South Africa  2197.033263  2.809246  2791243.75  0.570414  0.570064  
South Africa  2191.363362  2.857641  2846327.75  0.570888  0.554365  
South Africa  2191.363362  2.908202  2897706.00  0.570888  0.542998  

[608 rows x 10 columns]

In [6]:
#Choose the Year to Perform the Analysis
print(pwt_subset.groupby("year").count()) 

      countrycode  cgdpo  pop  emp  avh  hc  cn  labsh  ctfp
year                                                        
2010           61     61   61   61   61  61  61     61    61
2011           61     61   61   61   61  61  61     61    61
2012           61     61   61   61   61  61  61     61    61
2013           61     61   61   61   61  61  61     61    61
2014           60     60   60   60   60  60  60     60    60
2015           60     60   60   60   60  60  60     60    60
2016           61     61   61   61   61  61  61     61    61
2017           61     61   61   61   61  61  61     61    61
2018           61     61   61   61   61  61  61     61    61
2019           61     61   61   61   61  61  61     61    61


We will choose 2019: the latest year with the largets number of ebservations for our variables of interest.
*To Do: Robustness using 2010,2015, and an average for 2010-2019

In [9]:
#Descriptive Statistics for the Variables in our Dataset
print(pwt_subset[pwt_subset['year'] == 2019].describe([0.05,0.10,0.25,0.50,0.75,0.90,0.95]).round(2))

         year        cgdpo      pop     emp      avh     hc            cn  \
count    61.0        61.00    61.00   61.00    61.00  61.00  6.100000e+01   
mean   2019.0   1766051.39    86.23   40.07  1814.08   3.23  8.117018e+06   
std       0.0   3788319.57   252.13  120.07   240.46   0.43  1.636157e+07   
min    2019.0     17330.60     0.34    0.19  1380.61   2.17  6.244054e+04   
5%     2019.0     45079.35     0.87    0.46  1439.73   2.51  2.194441e+05   
10%    2019.0     71310.98     2.08    1.05  1505.56   2.70  3.189460e+05   
25%    2019.0    199241.38     5.53    2.85  1611.37   2.90  8.281802e+05   
50%    2019.0    496383.50    17.10    8.10  1791.74   3.27  2.566180e+06   
75%    2019.0   1862070.00    58.56   25.60  2019.92   3.59  8.721109e+06   
90%    2019.0   3120405.75   127.58   69.98  2142.10   3.72  1.885582e+07   
95%    2019.0   5031850.00   270.63  131.17  2168.92   3.77  2.610288e+07   
max    2019.0  20566034.00  1433.78  798.81  2330.17   4.35  1.015442e+08   

In [66]:
pwt_subset = pwt_subset[pwt_subset['year'] == 2019]

# Development Accounting

## GDP per worker

In [68]:
pwt_subset['GDP_pw'] = pwt_subset['cgdpo']/pwt_subset['emp']
pwt_subset['GDP_pw'].describe([0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95,0.99]).round(2)

count        61.00
mean      74177.78
std       36303.48
min       18371.85
1%        20269.11
5%        23559.62
10%       31758.38
25%       43949.43
50%       71907.51
75%       96217.70
90%      106398.89
95%      128644.07
99%      169521.02
max      219595.21
Name: GDP_pw, dtype: float64

In [69]:
log_var_y_pw = np.log(np.var(pwt_subset['GDP_pw']))
log_var_y_pw

20.9828084235369

In [70]:
pwt_subset['k']       = (pwt_subset['cn']/pwt_subset['emp'])
pwt_subset['y_hk_pw'] = (pwt_subset['k']**(1-pwt_subset['labsh']))*((pwt_subset['avh']*pwt_subset['hc'])**pwt_subset['labsh'])
pwt_subset['y_hk_pw'].describe([0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95,0.99]).round(2)

count        61.00
mean      38014.52
std       21809.12
min       13720.50
1%        15713.18
5%        19116.97
10%       21403.18
25%       27488.07
50%       34361.97
75%       42009.75
90%       53999.00
95%       58109.69
99%      126415.77
max      170276.42
Name: y_hk_pw, dtype: float64

In [71]:
log_var_y_hk_pw = np.log(np.var(pwt_subset['y_hk_pw']))
log_var_y_hk_pw

19.96363806709258

In [72]:
ypw_p99_p01_ratio    = pwt_subset['GDP_pw'].quantile(0.99)/pwt_subset['GDP_pw'].quantile(0.01)
yhk_pw_p99_p01_ratio = pwt_subset['y_hk_pw'].quantile(0.99)/pwt_subset['y_hk_pw'].quantile(0.01)
ypw_p99_p01_ratio,yhk_pw_p99_p01_ratio

(8.363515913075576, 8.045208034868505)

In [73]:
ypw_p95_p05_ratio    = pwt_subset['GDP_pw'].quantile(0.95)/pwt_subset['GDP_pw'].quantile(0.05)
yhk_pw_p95_p05_ratio = pwt_subset['y_hk_pw'].quantile(0.95)/pwt_subset['y_hk_pw'].quantile(0.05)
ypw_p95_p05_ratio,yhk_pw_p95_p05_ratio

(5.460361820923015, 3.039691074416934)

In [74]:
ypw_p90_p10_ratio    = pwt_subset['GDP_pw'].quantile(0.90)/pwt_subset['GDP_pw'].quantile(0.10)
yhk_pw_p90_p10_ratio = pwt_subset['y_hk_pw'].quantile(0.90)/pwt_subset['y_hk_pw'].quantile(0.10)
ypw_p90_p10_ratio,yhk_pw_p90_p10_ratio

(3.3502616739254916, 2.522942696685417)

In [75]:
ypw_p75_p25_ratio    = pwt_subset['GDP_pw'].quantile(0.75)/pwt_subset['GDP_pw'].quantile(0.25)
yhk_pw_p75_p25_ratio = pwt_subset['y_hk_pw'].quantile(0.75)/pwt_subset['y_hk_pw'].quantile(0.25)
ypw_p75_p25_ratio,yhk_pw_p75_p25_ratio

(2.189282284743125, 1.5282903739752431)

## GDP per worker: Measures of Success

In [83]:
success1_pw = log_var_y_hk_pw/log_var_y_pw
success1_pw

0.9514283152249015

In [84]:
success1_pw_99_01 = yhk_pw_p99_p01_ratio/ypw_p99_p01_ratio
success1_pw_99_01

0.9619409012291797

In [85]:
success1_pw_95_05 = yhk_pw_p95_p05_ratio/ypw_p95_p05_ratio
success1_pw_95_05

0.5566830869649416

In [86]:
success1_pw_90_10 = yhk_pw_p90_p10_ratio/ypw_p90_p10_ratio
success1_pw_90_10

0.7530584002799079

In [87]:
success1_pw_75_25 = yhk_pw_p75_p25_ratio/ypw_p75_p25_ratio
success1_pw_75_25

0.6980782627373984

Factors of production explain 95% of the variance in income per worker!

If we perform the analysis removing countries with very high or very low values (p99 and p01, p95 and p05, p90 and p10) we get that factors of production explain between 55% and 96% of cross country income differences. However, the measures of success are non-monotonic with respect to the inclusion of extreme values.

## Disecting the Measure of Success

###  1. Shuting Down the Variation in Labour Shares

We start by recalculating the measures of success but shitting down the variation in labour shares. We do two alternative calculations:

 * **Alternative 1**: Fix the labour share at 2/3, the most widely used value for in the literature
 * **Alternative 2**: Fix the labour share at the median of the pwt sample in 2019: 0.56

In [88]:
pwt_subset['y_hk_pw_alt1'] = (pwt_subset['cn']**(1/3))*((pwt_subset['emp']*pwt_subset['avh']*pwt_subset['hc'])**(2/3))
pwt_subset['y_hk_pw_alt1'] = pwt_subset['y_hk_pw_alt1']/pwt_subset['emp']

In [89]:
log_var_y_hk_pw_alt1 = np.log(np.var(pwt_subset['y_hk_pw_alt1']))
log_var_y_hk_pw_alt1

17.138250661223427

In [222]:
success1_pw_alt1 = log_var_y_hk_pw_alt1/log_var_y_pw

In [220]:
print("The Measure of Success Under *Alternative 1* is:") 
success1_pw_alt1.round(2)

The Measure of Success Under *Alternative 1* is:


0.82

In [226]:
0.82/0.95

0.8631578947368421

In [92]:
pwt_subset['y_hk_pw_alt2'] = (pwt_subset['cn']**(1-0.56))*((pwt_subset['emp']*pwt_subset['avh']*pwt_subset['hc'])**(0.56))
pwt_subset['y_hk_pw_alt2'] = pwt_subset['y_hk_pw_alt2']/pwt_subset['emp']

In [223]:
log_var_y_hk_pw_alt2 = np.log(np.var(pwt_subset['y_hk_pw_alt2']))
log_var_y_hk_pw_alt2.round(2)

18.37

In [224]:
success1_pw_alt2 = log_var_y_hk_pw_alt2/log_var_y_pw

In [225]:
print("The Measure of Success Under *Alternative 2* is:") 
success1_pw_alt2.round(2)

The Measure of Success Under *Alternative 2* is:


0.88

We choose **Alternative 1** to err on the conservative side. Shutting down the variability coming from differentials in factor intensities, we find that factors of production explain 81% of the variance in output per worker across countries.

### Variance Decomposition

To perform a full variance decomposition we need to calculate a few moments first.

In [185]:
var_log_y_hk     = np.var(np.log(pwt_subset['y_hk_pw_alt1']))
var_log_y_hk

0.06292096900033263

In [186]:
var_log_k     = np.var(np.log(pwt_subset['k']))

In [187]:
var_log_avh   = np.var(np.log(pwt_subset['avh']))

In [188]:
var_log_h     = np.var(np.log(pwt_subset['hc']))

In [189]:
cov_log_avh_h = np.cov(np.log(pwt_subset['avh']),np.log(pwt_subset['hc']))[0][1]

In [190]:
cov_log_k_h   = np.cov(np.log(pwt_subset['k']),np.log(pwt_subset['hc']))[0][1]

In [191]:
cov_log_k_avh = np.cov(np.log(pwt_subset['k']),np.log(pwt_subset['avh']))[0][1]

In [192]:
α             = 1/3

In [193]:
((α**2)*var_log_k)/var_log_y_hk

0.7610496429649978

In [194]:
(((1-α)**2)*var_log_avh)/var_log_y_hk

0.12417534709460554

In [195]:
(((1-α)**2)*var_log_h)/var_log_y_hk

0.13355037864726207

In [196]:
(2*((1-α)**2)*cov_log_avh_h)/var_log_y_hk

-0.0976768487221539

In [197]:
(2*α*(1-α)*cov_log_k_h)/var_log_y_hk

0.3843404157028929

In [198]:
(2*α*(1-α)*cov_log_k_avh)/var_log_y_hk

-0.30575185849938435

In [200]:
total_variation = np.abs(((α**2)*var_log_k)) + np.abs((((1-α)**2)*var_log_avh)) + np.abs((((1-α)**2)*var_log_h)) + np.abs((2*((1-α)**2)*cov_log_avh_h)) + np.abs((2*α*(1-α)*cov_log_k_h)) + np.abs((2*α*(1-α)*cov_log_k_avh))

In [218]:
total_variation.round(2)

0.11

In [203]:
#The share of the total variation given by capital per worker is

In [212]:
(np.abs(((α**2)*var_log_k))/total_variation).round(2)

0.42

In [213]:
#The share of the total variation given by hours worked is
(np.abs((((1-α)**2)*var_log_avh))/total_variation).round(2)

0.07

In [214]:
#The share of the total variation given by hours worked is
(np.abs((((1-α)**2)*var_log_h))/total_variation).round(2)

0.07

In [215]:
#The share of the total variation given by the interaction between hours worked and human capital
(np.abs((2*((1-α)**2)*cov_log_avh_h))/total_variation).round(2)

0.05

In [216]:
#The share of the total variation given by the interaction between capital and human capital
(np.abs((2*α*(1-α)*cov_log_k_h))/total_variation).round(2)

0.21

In [217]:
#The share of the total variation given by the interaction between capital and hours worked is
(np.abs((2*α*(1-α)*cov_log_k_avh))/total_variation).round(2)

0.17

## Conclusion

We perform a development accounting exercise ala Caselli, using PWT 10.0. We find that:

 * Factors of production explain 95% of the variation in income per worker
       * Out of that total variation, 86% is driven by pure factors of production, like capital, human capital, and hours worked, while the remaining 14% is explained by differences in factor shares across countries.
       * Within factors of production, differences in capital per worker explain 42% of the total variation, while human capital and hours account for 7% each.
       * We find that richer countries tend to have both higher levels of physical and human capital, with the joint comovement between these two factors of production explaining 21% of the share of income explained by factors only.
       * Moreover, rich countries tend to have higher levels of physical capital and lower levels of hours worked per worker, which reduces the total varition driven by factors of production, helping to explain 17% of the variation that is purely driven by productive factors
       * A remaining 5% of the total income differences explained by production factors is driven by the fact that in countries with higher levels of human capital workers tend to work fewer hours.